packages/excalidraw/change.ts

```diff
@@ -0,0 +1,1394 @@
+          flags.containsZindexDifference = true;
+
+          // As the element was force deleted, we need to check if adding it back results in a visible change
+          if (
+            partial.isDeleted === false ||
+            (partial.isDeleted !== true && element.isDeleted === false)
+          ) {
+            flags.containsVisibleDifference = true;
+          }
+        }
+      }
+
+      return element;
+    };
+
+  /**
+   * Check for visible changes regardless of whether they were removed, added or updated.
+   */
+  private static checkForVisibleDifference(
+    element: OrderedExcalidrawElement,
+    partial: ElementPartial,
+  ) {
+    if (element.isDeleted && partial.isDeleted !== false) {
+      // When it's deleted and partial is not false, it cannot end up with a visible change
+      return false;
+    }
+
+    if (element.isDeleted && partial.isDeleted === false) {
+      // When we add an element, it results in a visible change
+      return true;
+    }
+
+    if (element.isDeleted === false && partial.isDeleted) {
+      // When we remove an element, it results in a visible change
+      return true;
+    }
+
+    // Check for any difference on a visible element
+    return Delta.isRightDifferent(element, partial);
+  }
+
+  /**
+   * Precedes `fixTextBindings` by removing existing bound text element when adding a new one through history, so we won't end up with multiple.
+   *
+   * Note: performs mutation, which we might want to refactor away, not to end up in an inconsistent state.
+   */
+  private static unbindExistingTextElements(
+    boundElements: readonly BoundElement[],
+    elements: SceneElementsMap,
+  ): readonly BoundElement[] | null {
+    const boundTextElements = boundElements.filter((x) => x.type === "text");
+
+    for (const { id } of boundTextElements) {
+      const element = elements.get(id);
+
+      if (element) {
+        mutateElement(
+          element as ExcalidrawTextElement,
+          {
+            isDeleted: true,
+            containerId: null,
+          },
+          false,
+        );
+      }
+    }
+
+    // TODO: consider separating bound text into a separate property, so we would't have to do this
+    // https://github.com/excalidraw/excalidraw/pull/7348#discussion_r1523123549
+    const nextBoundTextElements = boundElements.filter(
+      (x) => x.type !== "text",
+    );
+
+    return nextBoundTextElements.length ? nextBoundTextElements : null;
+  }
+
+  /**
+   * Fixing text bindings so we won't end up in an incosistent state after applying changes.
+   *
+   * The logic is similar to one `fixBindingsAfterDeletion`, `repairBoundElement` and `repairContainerElement`, but serves a different purpose.
+   * While the above functions are designed to ensure consistent state during restore, these are designed to resolve conflicts in the least intrusive way during concurrent updates.
```

Comment:
Test once again current bindings behaviour and write down pros & cons, as i.e.:
- arrows will remain bound after element deletion through history, which could be confusing when trying rebind again
- elements will remain bound to frames after frame deletion through history, which could be confusing when trying to rebind to a different frame

Reply:
Likely does not make sense leaving it as is & will need to be fixed

Reply:
Fixed